### ISSUE TITLE
Incorrect Output in Model Inference Process

### ISSUE THREAD

**Issue by lalulxm** [YiyangZhou/LURE#8]
Hi! I have read your paper and it is really an amazing work! 
But I encountered some issues during the Model Inference stage. I followed the steps below for model inference.
1. Prepare the code and the environment
``` shell
git clone https://github.com/YiyangZhou/LURE.git
cd LURE
conda env create -f environment.yml
conda activate LURE
```
2. Download [Checkpoint Aligned with Vicuna 13B (stage 1）](https://drive.google.com/file/d/1u9FRRBB3VovP1HxCAlpD9Lw4t4P6-Yq8/view)
3. set the path to the pretrained checkpoint in the evaluation config file in [eval_configs/minigpt4_eval.yaml](https://github.com/YiyangZhou/LURE/blob/main/eval_configs/minigpt4_eval.yaml#L11)
4. Prepare model captions by running the following command: 
``` 
python output_LURE.py --mode == 'inference' --cfg-path /eval_configs/minigpt4_eval.yaml --gpu-id 0 --input_image /path/to/image_file --output_file /output/output.jsonl 
```
It's important to note that for each image, I modified the prompt to a specific question. Here are some of the output results. Some strange phenomena occurred here: **only part of the sentence was output for the caption.**
```
{"id": "64f51a50c1aff6f7.png", "question": "What color is the umbrella?", "caption": "The umbrella is purple, pink, and yellow.", "objs": ["umbrella", "yellow"], "plist": [0.393798828125], "p_all": {"The": [0.5888671875], "umbrella": [0.70556640625], "is": [0.404296875], "purple": [0.214111328125], ",": [], "pink": [0.1790771484375], "and": [0.568359375], "yellow": [0.393798828125], ".": []}, "model": "MiniGPT-4_13b"}
{"id": "01a4b22028ed8133.png", "question": "Describe the image in detail.", "caption": "The image is of a television with an old man in glasses and a", "objs": ["image", "television", "man", "glasses"], "plist": [0.374267578125, 0.384765625], "p_all": {"The": [0.61328125], "image": [0.6396484375], "is": [0.426025390625], "of": [0.27197265625], "a": [0.84814453125, 0.84814453125], "television": [0.374267578125], "with": [0.176513671875], "an": [0.1585693359375], "old": [0.19677734375], "man": [0.87646484375], "in": [0.412109375], "glasses": [0.384765625], "and": [0.58349609375]}, "model": "MiniGPT-4_13b"}
{"id": "856c000622d66bd9.png", "question": "Describe the image in detail.", "caption": "The image shows a row of shops on a", "objs": ["image", "row", "shops"], "plist": [0.10723876953125], "p_all": {"The": [0.4541015625], "image": [0.7734375], "shows": [0.56005859375], "a": [0.86865234375, 0.86865234375], "row": [0.10723876953125], "of": [1.0], "shops": [0.7548828125], "on": [0.2744140625]}, "model": "MiniGPT-4_13b"}
{"id": "50ba2afa85a784cb.png", "question": "In the image, which truck (left or right) has its door open?", "caption": "The red truck on the left has its door open.\n\n </Img>", "objs": ["truck", "left", "door", "/Img", ">"], "plist": [0.355712890625], "p_all": {"The": [0.406494140625], "red": [0.3154296875], "truck": [0.355712890625], "on": [0.57861328125], "the": [1.0], "left": [0.80078125], "has": [1.0], "its": [1.0], "door": [1.0], "open": [1.0], ".": [], "<": [], "/Img": [], ">": []}, "model": "MiniGPT-4_13b"}
{"id": "071af503f95aee59.png", "question": "How many oysters can you see in the photo?", "caption": "I can see 5 oysters on the plate", "objs": ["oysters", "plate"], "plist": [], "p_all": {"I": [0.368896484375], "can": [0.52734375], "see": [1.0], "5": [0.1383056640625], "oysters": [1.0], "on": [0.4609375], "the": [1.0], "plate": [1.0]}, "model": "MiniGPT-4_13b"}
```
5.  get the masked caption of prepared captions
```
python generate_IDK.py   --input_file /output/output.jsonl  --output_file /output/idk_caption_file.jsonl
```
The following are the corresponding outputs in JSON format for the 5 entries:
```
{"id": "64f51a50c1aff6f7.png", "question": "What color is the umbrella?", "caption": "The umbrella is purple, pink, [IDK].", "objs": ["umbrella", "yellow"], "plist": [0.393798828125], "p_all": {"The": [0.5888671875], "umbrella": [0.70556640625], "is": [0.404296875], "purple": [0.214111328125], ",": [], "pink": [0.1790771484375], "and": [0.568359375], "yellow": [0.393798828125], ".": []}, "model": "MiniGPT-4_13b"}
{"id": "01a4b22028ed8133.png", "question": "Describe the image in detail.", "caption": "The image is of [IDK] an old man [IDK] a", "objs": ["image", "television", "man", "glasses"], "plist": [0.374267578125, 0.384765625], "p_all": {"The": [0.61328125], "image": [0.6396484375], "is": [0.426025390625], "of": [0.27197265625], "a": [0.84814453125, 0.84814453125], "television": [0.374267578125], "with": [0.176513671875], "an": [0.1585693359375], "old": [0.19677734375], "man": [0.87646484375], "in": [0.412109375], "glasses": [0.384765625], "and": [0.58349609375]}, "model": "MiniGPT-4_13b"}
{"id": "856c000622d66bd9.png", "question": "Describe the image in detail.", "caption": "The image shows [IDK] shops on a", "objs": ["image", "row", "shops"], "plist": [0.10723876953125], "p_all": {"The": [0.4541015625], "image": [0.7734375], "shows": [0.56005859375], "a": [0.86865234375, 0.86865234375], "row": [0.10723876953125], "of": [1.0], "shops": [0.7548828125], "on": [0.2744140625]}, "model": "MiniGPT-4_13b"}
{"id": "50ba2afa85a784cb.png", "question": "In the image, which truck (left or right) has its door open?", "caption": "The [IDK] the left has its door open. </Img>", "objs": ["truck", "left", "door", "/Img", ">"], "plist": [0.355712890625], "p_all": {"The": [0.406494140625], "red": [0.3154296875], "truck": [0.355712890625], "on": [0.57861328125], "the": [1.0], "left": [0.80078125], "has": [1.0], "its": [1.0], "door": [1.0], "open": [1.0], ".": [], "<": [], "/Img": [], ">": []}, "model": "MiniGPT-4_13b"}
{"id": "071af503f95aee59.png", "question": "How many oysters can you see in the photo?", "caption": "I can see 5 oysters on [IDK]", "objs": ["oysters", "plate"], "plist": [], "p_all": {"I": [0.368896484375], "can": [0.52734375], "see": [1.0], "5": [0.1383056640625], "oysters": [1.0], "on": [0.4609375], "the": [1.0], "plate": [1.0]}, "model": "MiniGPT-4_13b"}
```
6.  obtain the rewriting response
```
python output_LURE.py --mode == 'rewrite' --cfg-path /eval_configs/minigpt4_eval.yaml --gpu-id 0 --input_caption /output/idk_caption_file  --input_image /path/to/image_file --output_file /output/answer.jsonl 
```
The final model output results are as follows: 
there are more errors in the captions, for example, **only outputting a** `\u200b` or **prompts like** `can't see the image`, etc.
```
{"id": "64f51a50c1aff6f7.png", "question": "According to the picture, remove the information that does not exist in the following description: The umbrella is purple, pink, [IDK].", "caption": "Here is the corrected image description:", "model": "LURE"}
{"id": "01a4b22028ed8133.png", "question": "According to the picture, remove the information that does not exist in the following description: The image is of [IDK] an old man [IDK] a", "caption": "\u200b", "model": "LURE"}
{"id": "856c000622d66bd9.png", "question": "According to the picture, remove the information that does not exist in the following description: The image shows [IDK] shops on a", "caption": "autumn day, which is well kept with the colours of autumn leaves", "model": "LURE"}
{"id": "50ba2afa85a784cb.png", "question": "According to the picture, remove the information that does not exist in the following description: The [IDK] the left has its door open. </Img>", "caption": "I'm sorry, I don't see the image you provided. Can you please try again or provide more context?", "model": "LURE"}
{"id": "071af503f95aee59.png", "question": "According to the picture, remove the information that does not exist in the following description: I can see 5 oysters on [IDK]", "caption": "What type of food is on the plate?", "model": "LURE"}
```
So, which step in my model inference process went wrong, leading to the incorrect output?

**Comment by YiyangZhou** [comment]
> Hi! I have read your paper and it is really an amazing work! But I encountered some issues during the Model Inference stage. I followed the steps below for model inference.
> 
> 1. Prepare the code and the environment
> 
> ```shell
> git clone https://github.com/YiyangZhou/LURE.git
> cd LURE
> conda env create -f environment.yml
> conda activate LURE
> ```
> 
> 2. Download [Checkpoint Aligned with Vicuna 13B (stage 1）](https://drive.google.com/file/d/1u9FRRBB3VovP1HxCAlpD9Lw4t4P6-Yq8/view)
> 3. set the path to the pretrained checkpoint in the evaluation config file in [eval_configs/minigpt4_eval.yaml](https://github.com/YiyangZhou/LURE/blob/main/eval_configs/minigpt4_eval.yaml#L11)
> 4. Prepare model captions by running the following command:
> 
> ```
> python output_LURE.py --mode == 'inference' --cfg-path /eval_configs/minigpt4_eval.yaml --gpu-id 0 --input_image /path/to/image_file --output_file /output/output.jsonl 
> ```
> 
> It's important to note that for each image, I modified the prompt to a specific question. Here are some of the output results. Some strange phenomena occurred here: **only part of the sentence was output for the caption.**
> 
> ```
> {"id": "64f51a50c1aff6f7.png", "question": "What color is the umbrella?", "caption": "The umbrella is purple, pink, and yellow.", "objs": ["umbrella", "yellow"], "plist": [0.393798828125], "p_all": {"The": [0.5888671875], "umbrella": [0.70556640625], "is": [0.404296875], "purple": [0.214111328125], ",": [], "pink": [0.1790771484375], "and": [0.568359375], "yellow": [0.393798828125], ".": []}, "model": "MiniGPT-4_13b"}
> {"id": "01a4b22028ed8133.png", "question": "Describe the image in detail.", "caption": "The image is of a television with an old man in glasses and a", "objs": ["image", "television", "man", "glasses"], "plist": [0.374267578125, 0.384765625], "p_all": {"The": [0.61328125], "image": [0.6396484375], "is": [0.426025390625], "of": [0.27197265625], "a": [0.84814453125, 0.84814453125], "television": [0.374267578125], "with": [0.176513671875], "an": [0.1585693359375], "old": [0.19677734375], "man": [0.87646484375], "in": [0.412109375], "glasses": [0.384765625], "and": [0.58349609375]}, "model": "MiniGPT-4_13b"}
> {"id": "856c000622d66bd9.png", "question": "Describe the image in detail.", "caption": "The image shows a row of shops on a", "objs": ["image", "row", "shops"], "plist": [0.10723876953125], "p_all": {"The": [0.4541015625], "image": [0.7734375], "shows": [0.56005859375], "a": [0.86865234375, 0.86865234375], "row": [0.10723876953125], "of": [1.0], "shops": [0.7548828125], "on": [0.2744140625]}, "model": "MiniGPT-4_13b"}
> {"id": "50ba2afa85a784cb.png", "question": "In the image, which truck (left or right) has its door open?", "caption": "The red truck on the left has its door open.\n\n </Img>", "objs": ["truck", "left", "door", "/Img", ">"], "plist": [0.355712890625], "p_all": {"The": [0.406494140625], "red": [0.3154296875], "truck": [0.355712890625], "on": [0.57861328125], "the": [1.0], "left": [0.80078125], "has": [1.0], "its": [1.0], "door": [1.0], "open": [1.0], ".": [], "<": [], "/Img": [], ">": []}, "model": "MiniGPT-4_13b"}
> {"id": "071af503f95aee59.png", "question": "How many oysters can you see in the photo?", "caption": "I can see 5 oysters on the plate", "objs": ["oysters", "plate"], "plist": [], "p_all": {"I": [0.368896484375], "can": [0.52734375], "see": [1.0], "5": [0.1383056640625], "oysters": [1.0], "on": [0.4609375], "the": [1.0], "plate": [1.0]}, "model": "MiniGPT-4_13b"}
> ```
> 
> 5. get the masked caption of prepared captions
> 
> ```
> python generate_IDK.py   --input_file /output/output.jsonl  --output_file /output/idk_caption_file.jsonl
> ```
> 
> The following are the corresponding outputs in JSON format for the 5 entries:
> 
> ```
> {"id": "64f51a50c1aff6f7.png", "question": "What color is the umbrella?", "caption": "The umbrella is purple, pink, [IDK].", "objs": ["umbrella", "yellow"], "plist": [0.393798828125], "p_all": {"The": [0.5888671875], "umbrella": [0.70556640625], "is": [0.404296875], "purple": [0.214111328125], ",": [], "pink": [0.1790771484375], "and": [0.568359375], "yellow": [0.393798828125], ".": []}, "model": "MiniGPT-4_13b"}
> {"id": "01a4b22028ed8133.png", "question": "Describe the image in detail.", "caption": "The image is of [IDK] an old man [IDK] a", "objs": ["image", "television", "man", "glasses"], "plist": [0.374267578125, 0.384765625], "p_all": {"The": [0.61328125], "image": [0.6396484375], "is": [0.426025390625], "of": [0.27197265625], "a": [0.84814453125, 0.84814453125], "television": [0.374267578125], "with": [0.176513671875], "an": [0.1585693359375], "old": [0.19677734375], "man": [0.87646484375], "in": [0.412109375], "glasses": [0.384765625], "and": [0.58349609375]}, "model": "MiniGPT-4_13b"}
> {"id": "856c000622d66bd9.png", "question": "Describe the image in detail.", "caption": "The image shows [IDK] shops on a", "objs": ["image", "row", "shops"], "plist": [0.10723876953125], "p_all": {"The": [0.4541015625], "image": [0.7734375], "shows": [0.56005859375], "a": [0.86865234375, 0.86865234375], "row": [0.10723876953125], "of": [1.0], "shops": [0.7548828125], "on": [0.2744140625]}, "model": "MiniGPT-4_13b"}
> {"id": "50ba2afa85a784cb.png", "question": "In the image, which truck (left or right) has its door open?", "caption": "The [IDK] the left has its door open. </Img>", "objs": ["truck", "left", "door", "/Img", ">"], "plist": [0.355712890625], "p_all": {"The": [0.406494140625], "red": [0.3154296875], "truck": [0.355712890625], "on": [0.57861328125], "the": [1.0], "left": [0.80078125], "has": [1.0], "its": [1.0], "door": [1.0], "open": [1.0], ".": [], "<": [], "/Img": [], ">": []}, "model": "MiniGPT-4_13b"}
> {"id": "071af503f95aee59.png", "question": "How many oysters can you see in the photo?", "caption": "I can see 5 oysters on [IDK]", "objs": ["oysters", "plate"], "plist": [], "p_all": {"I": [0.368896484375], "can": [0.52734375], "see": [1.0], "5": [0.1383056640625], "oysters": [1.0], "on": [0.4609375], "the": [1.0], "plate": [1.0]}, "model": "MiniGPT-4_13b"}
> ```
> 
> 6. obtain the rewriting response
> 
> ```
> python output_LURE.py --mode == 'rewrite' --cfg-path /eval_configs/minigpt4_eval.yaml --gpu-id 0 --input_caption /output/idk_caption_file  --input_image /path/to/image_file --output_file /output/answer.jsonl 
> ```
> 
> The final model output results are as follows: there are more errors in the captions, for example, **only outputting a** `\u200b` or **prompts like** `can't see the image`, etc.
> 
> ```
> {"id": "64f51a50c1aff6f7.png", "question": "According to the picture, remove the information that does not exist in the following description: The umbrella is purple, pink, [IDK].", "caption": "Here is the corrected image description:", "model": "LURE"}
> {"id": "01a4b22028ed8133.png", "question": "According to the picture, remove the information that does not exist in the following description: The image is of [IDK] an old man [IDK] a", "caption": "\u200b", "model": "LURE"}
> {"id": "856c000622d66bd9.png", "question": "According to the picture, remove the information that does not exist in the following description: The image shows [IDK] shops on a", "caption": "autumn day, which is well kept with the colours of autumn leaves", "model": "LURE"}
> {"id": "50ba2afa85a784cb.png", "question": "According to the picture, remove the information that does not exist in the following description: The [IDK] the left has its door open. </Img>", "caption": "I'm sorry, I don't see the image you provided. Can you please try again or provide more context?", "model": "LURE"}
> {"id": "071af503f95aee59.png", "question": "According to the picture, remove the information that does not exist in the following description: I can see 5 oysters on [IDK]", "caption": "What type of food is on the plate?", "model": "LURE"}
> ```
> 
> So, which step in my model inference process went wrong, leading to the incorrect output?

Hello, I'm very sorry for the late response. I noticed that you mentioned downloading the "Checkpoint Aligned with Vicuna 13B (stage 1)" provided by Minigpt-4. This checkpoint has only undergone pre-training, and its responses to questions are discontinuous. If you need to obtain normal responses, you can download the Minigpt-4 checkpoint version that has undergone instruction tuning through this link: https://drive.google.com/file/d/1a4zLvaiDBr-36pasffmgpvH5P7CKmpze/view
